### PR TITLE
Starter overview documentation and guidelines

### LINKED ISSUE #16: Document custom stream application guidelines
As a user, I'd like to refer to docs when developing custom `source`, `processor`, or `sink` applications, so I can develop by following the best practices.

**Acceptance:**
- reference guide include guidelines for each application type
- unit-testing section is included 
- troubleshooting techniques listed
- sample applications available for reference 

This could possibly evolve from [this](https://docs.google.com/document/d/1fAh4S-rN0i-lVwr0pIFE2L52CfQ2WZ7mKO-WiQ2VHtA/edit#heading=h.d68zgr59juff) document.

### PR DESCRIPTION
**Pull request by mbogoevici**
Fixes #16

### CODE REVIEW

File: spring-cloud-stream-app-starters-docs/src/main/asciidoc/overview.adoc [hunk 1]
```diff
@@ -1,8 +1,185 @@
 [[overview]]
-== Spring Cloud Stream Modules 
 
-This section goes into more detail about how you can work with Spring Cloud Stream Module as standalone applications or with Spring Cloud Data Flow.
+This section will provide you with a detailed overview of Spring Cloud Stream Application Starters, their purpose, and how to use them.
+It assumes familiarity with general Spring Cloud Stream concepts, which can be found in its reference documentation.
```
Comment (mbogoevici) [inline spring-cloud-stream-app-starters-docs/src/main/asciidoc/overview.adoc]: Add link to SCSt reference documentation.

File: spring-cloud-stream-app-starters-docs/src/main/asciidoc/overview.adoc [hunk 2]
```diff
@@ -1,8 +1,185 @@
...
 
-=== Overview
+== Introduction
 
-TBD
+Spring Cloud Stream Application Starters provide you with predefined Spring Cloud Stream applications that you can run independently or with Spring Cloud Data Flow.
```
Comment (mbogoevici) [inline spring-cloud-stream-app-starters-docs/src/main/asciidoc/overview.adoc:8]: Add link to SCDF reference documentation.

**Comment by fifthposition** [comment]
(Pardon the dumb question please...) Is a goal to expand on some of these sections (e.g. “Creating your own starter”)?

File: spring-cloud-stream-app-starters-docs/src/main/asciidoc/overview.adoc (hunk 1 above)
Comment (fifthposition) [inline spring-cloud-stream-app-starters-docs/src/main/asciidoc/overview.adoc]: “can be found in the Spring Cloud Stream reference documentation”

File: spring-cloud-stream-app-starters-docs/src/main/asciidoc/overview.adoc (hunk 2 above)
Comment (fifthposition) [inline spring-cloud-stream-app-starters-docs/src/main/asciidoc/overview.adoc:8]: Should “Application Starters” be “application starters” (non-title)? (Also: “Data Flow” or “Dataflow”? I’d gotten used to the one-word version. 😃)

File: spring-cloud-stream-app-starters-docs/src/main/asciidoc/overview.adoc [hunk 3]
```diff
@@ -1,8 +1,185 @@
...
+You can also use the starters as basis for creating your own applications.
```
Comment (fifthposition) [inline spring-cloud-stream-app-starters-docs/src/main/asciidoc/overview.adoc]: “a basis for”

File: spring-cloud-stream-app-starters-docs/src/main/asciidoc/overview.adoc [hunk 4]
```diff
@@ -1,8 +1,185 @@
...
+They include:
+
+* connectors (sources and sinks) for middleware including message brokers, storage (relational, non-relational, filesystem);
+* adapters for various network protocols;
+* generic processors that can be customized via SpEL or scripting.
```
Comment (fifthposition) [inline spring-cloud-stream-app-starters-docs/src/main/asciidoc/overview.adoc]: Link to [SpEL docs](https://docs.spring.io/spring/docs/current/spring-framework-reference/html/expressions.html)?

File: spring-cloud-stream-app-starters-docs/src/main/asciidoc/overview.adoc [hunk 5]
```diff
@@ -1,8 +1,185 @@
...
+
+You can find a detailed listing of all the starters, as well as their options in the <<starters,corresponding>> section of this guide.
```
Comment (fifthposition) [inline spring-cloud-stream-app-starters-docs/src/main/asciidoc/overview.adoc]: “all of the starters and their options in the… section”

File: spring-cloud-stream-app-starters-docs/src/main/asciidoc/overview.adoc [hunk 6]
```diff
@@ -1,8 +1,185 @@
...
+
+=== Starters and pre-built applications
+
+As a user of Spring Cloud Stream Application Starters you have access to two types of artifacts.
+
+_Starters_ are libraries that contain the complete configuration of a Spring Cloud Stream application with a specific role (e.g. an _http source_ that receives HTTP POST requests and forwards the data on its output channel to downstream Spring Cloud Stream applications).
```
Comment (fifthposition) [inline spring-cloud-stream-app-starters-docs/src/main/asciidoc/overview.adoc]: “e.g., an _HTTP source_”

File: spring-cloud-stream-app-starters-docs/src/main/asciidoc/overview.adoc [hunk 7]
```diff
@@ -1,8 +1,185 @@
...
+Starters are not executable applications, and are intended be included in other Spring Boot applications, along with a Binder implementation.
```
Comment (fifthposition) [inline spring-cloud-stream-app-starters-docs/src/main/asciidoc/overview.adoc]: “intended to be included”

File: spring-cloud-stream-app-starters-docs/src/main/asciidoc/overview.adoc [hunk 8]
```diff
@@ -1,8 +1,185 @@
...
+
+_Prebuilt applications_ are Spring Boot applications that include the starters and a binder implementation.
```
Comment (fifthposition) [inline spring-cloud-stream-app-starters-docs/src/main/asciidoc/overview.adoc]: “Binder” or “binder”? (I am actually not sure. 😃)

File: spring-cloud-stream-app-starters-docs/src/main/asciidoc/overview.adoc [hunk 9]
```diff
@@ -1,8 +1,185 @@
...
+Prebuilt applications are uberjars and include minimal code required to execute standalone.
```
Comment (fifthposition) [inline spring-cloud-stream-app-starters-docs/src/main/asciidoc/overview.adoc]: “include the minimal code necessary for standalone execution.”

Could there be an explanation of “uberjars” (an “i.e.” or something)?

File: spring-cloud-stream-app-starters-docs/src/main/asciidoc/overview.adoc [hunk 10]
```diff
@@ -1,8 +1,185 @@
...
+For each starter, the project provides a prebuilt version including the Kafka binder, as well as one including the Rabbit MQ binder.
```
Comment (fifthposition) [inline spring-cloud-stream-app-starters-docs/src/main/asciidoc/overview.adoc]: “For each starter, the project provides a prebuilt version including the Kafka binder and a prebuilt version including the RabbitMQ binder.”

File: spring-cloud-stream-app-starters-docs/src/main/asciidoc/overview.adoc [hunk 11]
```diff
@@ -1,8 +1,185 @@
...
+
+[NOTE]
+Only starters are present in the source code of the project.
+Prebuilt applications are generated according to the maven plugin configuration.
```
Comment (fifthposition) [inline spring-cloud-stream-app-starters-docs/src/main/asciidoc/overview.adoc]: “Maven plugin”

File: spring-cloud-stream-app-starters-docs/src/main/asciidoc/overview.adoc [hunk 12]
```diff
@@ -1,8 +1,185 @@
...
+
+The prebuilt applications follow a naming convention too: `<functionality>-<type>-<binder>`. For example, `cassandra-sink-kafka` is a _Cassandra sink_ using the Kafka binder.
+
+=== Using the artifacts
+
+You either get access to the artifacts produced by Spring Cloud Stream Application Starters via Maven, Docker, or building the artifacts yourself.
+
+==== Maven and Docker accesss
+
+Starters are available as Maven artifacts in the [Spring repositories](). You can add them as dependencies to your application, as follows:
```
Comment (fifthposition) [inline spring-cloud-stream-app-starters-docs/src/main/asciidoc/overview.adoc]: No link.

File: spring-cloud-stream-app-starters-docs/src/main/asciidoc/overview.adoc [hunk 13]
```diff
@@ -1,8 +1,185 @@
...
+
+First, you need to generate the prebuilt applications.
+This is done by running the application generation Maven plugin.
+You can do so by simply invoking the corresponding script in the root of the project.
+
+```
+./generate.sh
+```
+
+The script will generate for the each of the prebuilt applications the following items:
```
Comment (fifthposition) [inline spring-cloud-stream-app-starters-docs/src/main/asciidoc/overview.adoc]: “For each of the prebuilt applications, the script will generate the following items:”

File: spring-cloud-stream-app-starters-docs/src/main/asciidoc/overview.adoc [hunk 14]
```diff
@@ -1,8 +1,185 @@
...
+* `pom.xml` file with the required dependencies (starter and binder)
+* a class that contains the `main` method of the application and imports the predefined configuration
+* generated integration test code that exercises the component against the configured binder.
+
+For example, `spring-cloud-starter-stream-sink-cassandra` will generate `cassandra-sink-rabbit` and `cassandra-sink-kafka` as completely functional applications.
+
+=== Creating custom artifacts
+
+Apart from accessing the sources, sinks and processors already provided by the project.
+In this section we will describe how to:
```
Comment (fifthposition) [inline spring-cloud-stream-app-starters-docs/src/main/asciidoc/overview.adoc]: “Apart from accessing the sources, sinks and processors already provided by the project, in this section we will describe how to:”

File: spring-cloud-stream-app-starters-docs/src/main/asciidoc/overview.adoc [hunk 15]
```diff
@@ -1,8 +1,185 @@
...
+
+	public static void main(String[] args) {
+		SpringApplication.run(CassandraSinkGemfireApplication.class, args);
+	}
+}
+----
+
+==== Creating your own applications and starters
+
+Spring Cloud Stream Application Starters consists of regular Spring Cloud Stream applications, with some additional conventions that facilitates generating prebuilt applications with the preconfigured binders.
```
Comment (fifthposition) [inline spring-cloud-stream-app-starters-docs/src/main/asciidoc/overview.adoc]: “Spring Cloud Stream Application Starters consists of regular Spring Cloud Stream applications with some additional conventions that facilitate generating prebuilt applications with the preconfigured binders.”

File: spring-cloud-stream-app-starters-docs/src/main/asciidoc/overview.adoc (hunk 2 above)
Comment (mbogoevici) [inline spring-cloud-stream-app-starters-docs/src/main/asciidoc/overview.adoc:8]: The official name of the project is "Spring Cloud Data Flow" (even if we use both forms).

File: spring-cloud-stream-app-starters-docs/src/main/asciidoc/overview.adoc (hunk 15 above)
Comment (sobychacko) [inline spring-cloud-stream-app-starters-docs/src/main/asciidoc/overview.adoc]: Shall we mention the fact that if anyone wants to build the docker images locally, they can do that? by invoking the docker:build goal on the generated artifacts. If for any reason, someone wants to run the image in a local docker container.

**Comment by fifthposition** [comment]
Rebased, squashed, and merged as https://github.com/spring-cloud/spring-cloud-stream-app-starters/commit/6d245d39bf3acd19b25fdda85d74cad41ddb639a.

**Pull request closed without merging**